controller_manager:
  ros__parameters:
    update_rate: 1000

    dynaarm_status_broadcaster:
      type: dynaarm_controllers/StatusBroadcaster

    gravity_compensation_controller:
      type: dynaarm_controllers/GravityCompensationController

    freedrive_controller:
      type: dynaarm_controllers/FreeDriveController

    position_controller:
      type: forward_command_controller/ForwardCommandController

    joint_state_broadcaster:
      type: joint_state_broadcaster/JointStateBroadcaster

    joint_trajectory_controller:
      type: joint_trajectory_controller/JointTrajectoryController

    effort_controller:
      type: effort_controllers/JointGroupEffortController

    cartesian_motion_controller:
      type: cartesian_motion_controller/CartesianMotionController

    motion_control_handle:
      type: cartesian_controller_handles/MotionControlHandle

    freeze_controller:
      type: dynaarm_controllers/FreezeController

    pid_tuner:
      type: dynaarm_controllers/PIDTuner

pid_tuner:
  ros__parameters:
    joints:
      - shoulder_rotation
      - shoulder_flexion
      - elbow_flexion
      - forearm_rotation
      - wrist_flexion
      - wrist_rotation

dynaarm_status_broadcaster:
  ros__parameters:
    joints:
      - shoulder_rotation
      - shoulder_flexion
      - elbow_flexion
      - forearm_rotation
      - wrist_flexion
      - wrist_rotation

motion_control_handle:
  ros__parameters:
    end_effector_link: "flange"
    robot_base_link: "base"
    joints:
      - shoulder_rotation
      - shoulder_flexion
      - elbow_flexion
      - forearm_rotation
      - wrist_flexion
      - wrist_rotation

cartesian_motion_controller:
  ros__parameters:
    end_effector_link: "flange"
    robot_base_link: "base"
    joints:
      - shoulder_rotation
      - shoulder_flexion
      - elbow_flexion
      - forearm_rotation
      - wrist_flexion
      - wrist_rotation
    command_interfaces:
      - position
      #- velocity
    solver:
        error_scale: 1.0
        iterations: 10
    pd_gains:
        trans_x: {p: 1.0}
        trans_y: {p: 1.0}
        trans_z: {p: 1.0}
        rot_x: {p: 0.5}
        rot_y: {p: 0.5}
        rot_z: {p: 0.5}

joint_trajectory_controller:
  ros__parameters:
    joints:
      - shoulder_rotation
      - shoulder_flexion
      - elbow_flexion
      - forearm_rotation
      - wrist_flexion
      - wrist_rotation
    command_interfaces:
      - position
      - velocity
    state_interfaces:
      - position
      - velocity
    # See https://control.ros.org/jazzy/doc/ros2_controllers/joint_trajectory_controller/doc/parameters.html for a full list of all parameters
    allow_partial_joints_goal: false # trajectory needs to contains goals from all joints handled by the controller
    constraints:
      stopped_velocity_tolerance: 0.5
      goal_time: 0.05
      # Tolerances for each joint
      shoulder_rotation: { trajectory: 0.5, goal: 0.05 }
      shoulder_flexion: { trajectory: 0.5, goal: 0.05 }
      elbow_flexion: { trajectory: 0.5, goal: 0.05 }
      forearm_rotation: { trajectory: 0.5, goal: 0.05 }
      wrist_flexion: { trajectory: 0.5, goal: 0.05 }
      wrist_rotation: { trajectory: 0.5, goal: 0.05 }

position_controller:
  ros__parameters:
    joints:
      - shoulder_rotation
      - shoulder_flexion
      - elbow_flexion
      - forearm_rotation
      - wrist_flexion
      - wrist_rotation
    interface_name: position

effort_controller:
  ros__parameters:
    joints:
      - shoulder_rotation
      - shoulder_flexion
      - elbow_flexion
      - forearm_rotation
      - wrist_flexion
      - wrist_rotation

gravity_compensation_controller:
  ros__parameters:
    joints:
      - shoulder_rotation
      - shoulder_flexion
      - elbow_flexion
      - forearm_rotation
      - wrist_flexion
      - wrist_rotation

freedrive_controller:
  ros__parameters:
    joints:
      - shoulder_rotation
      - shoulder_flexion
      - elbow_flexion
      - forearm_rotation
      - wrist_flexion
      - wrist_rotation
